spec/controllers/saml_idp_controller_spec.rb

```diff
@@ -296,6 +296,58 @@
           expect(@analytics).to have_received(:track_event).
             with(Analytics::SAML_AUTH, analytics_hash)
         end
+
+        it 'returns the requested AAL level defined by sp' do
+          stub_analytics
+          allow(@analytics).to receive(:track_event)
+
+          user = create(:user, :signed_up)
+          auth_settings = requested_aal2_authn_context_saml_settings
+          IdentityLinker.new(user, auth_settings.issuer).link_identity
+          user.identities.last.update!(verified_attributes: ['email'])
+          generate_saml_response(user, auth_settings)
+
+          expect(response.status).to eq(200)
+
+          analytics_hash = {
+            success: true,
+            errors: {},
+            nameid_format: Saml::Idp::Constants::NAME_ID_FORMAT_PERSISTENT,
+            authn_context: [Saml::Idp::Constants::AAL2_AUTHN_CONTEXT_CLASSREF],
+            service_provider: 'http://localhost:3000',
+            idv: false,
+            finish_profile: false,
+          }
+
+          expect(@analytics).to have_received(:track_event).
+            with(Analytics::SAML_AUTH, analytics_hash)
+        end
+      end
+
+      it 'returns the requested IAL level defined by sp' do
+        stub_analytics
+        allow(@analytics).to receive(:track_event)
+
+        user = create(:user, :signed_up)
+        auth_settings = requested_ial1_authn_context_saml_settings
+        IdentityLinker.new(user, auth_settings.issuer).link_identity
+        user.identities.last.update!(verified_attributes: ['email'])
+        generate_saml_response(user, auth_settings)
+
+        expect(response.status).to eq(200)
+
+        analytics_hash = {
+          success: true,
+          errors: {},
+          nameid_format: Saml::Idp::Constants::NAME_ID_FORMAT_PERSISTENT,
+          authn_context: [Saml::Idp::Constants::IAL1_AUTHN_CONTEXT_CLASSREF],
+          service_provider: 'http://localhost:3000',
+          idv: false,
+          finish_profile: false,
+        }
+
+        expect(@analytics).to have_received(:track_event).
+          with(Analytics::SAML_AUTH, analytics_hash)
```

Comment:
These expectations don't make much sense to me?

I would expect us to check the output of the XML response for the correct authn_context? Add some methods to `SamlResponseDoc` to make it easier? Something like:

```
expect(xmldoc.authn_context).to eq(IAL1_AUTHN_CONTEXT_CLASSREF)
```

Reply:
Thanks for the suggestion. We are now checking the output of the XML response for the correct authn_context. I cleaned up other authn_context tests as well.